### ISSUE TITLE
Proper way to handle blocking C functions in threaded builds?

### ISSUE THREAD

**Issue by ecraven** [cisco/ChezScheme#277]
I'm having a bit of trouble wrapping C functions like `recv` and `send` which can block for arbitrarily long times.

I'm trying to wrap them using *only* Scheme code (and the FFI), no C code. I've been successful in non-threaded builds (or threaded builds with only a single thread), but I'm seeing strange blocking behaviour of other threads in threaded builds.

I have code similar to
````
(lock-object buffer)
((foreign-procedure "recv" ...) buffer ...)
(unlock-object buffer)
````
From reading csug, I think I might have to call `Sactivate_thread` and `Sdeactivate_thread` somewhere here, but can I even do that from inside Scheme? I want to avoid building and linking C code at all costs.

Thanks for any assistance with this or clarifications on what I'm doing wrong :)

**Comment by evilbinary** [comment]
You could do this  https://github.com/evilbinary/scheme-lib/blob/master/packages/cffi/cffi.ss wrap Sdeactivate_thread  and Sactivate_thread. cffi is a usefull  lib to call c with thread.

**Comment by dybvig** [comment]
You should not run any Scheme code or touch any unlocked Scheme objects after deactivating and before reactivating.  If I understand the proposed solution, it would require calling `Sdeactivate_thread`, and returning to Scheme to call the blocking foreign procedure, which would violate this restriction.

I believe pull request #267 is intended to address this issue.

**Comment by ecraven** [comment]
I'll close this, it should be solved if #267 is merged.
